File: Robotics_Kinematics_and_Dynamics_Course/Chapter17/Lesson1/Chapter17_Lesson1.java
import org.ejml.simple.SimpleMatrix;

public class FloatingBaseMomentum {
    public static void main(String[] args) {
        int nJoints = 12;
        int ndof = 6 + nJoints;

        // Generalized coordinates q and velocities v
        SimpleMatrix q = new SimpleMatrix(ndof, 1);
        SimpleMatrix v = new SimpleMatrix(ndof, 1);

        // Example inertia matrix H(q); identity for demonstration
        SimpleMatrix H = SimpleMatrix.identity(ndof);

        // Set example base twist (omega_x, omega_y, omega_z, v_x, v_y, v_z)
        v.set(0, 0.0);
        v.set(1, 0.0);
        v.set(2, 0.5);
        v.set(3, 0.1);
        v.set(4, 0.0);
        v.set(5, 0.0);

        // Set joint velocities
        for (int i = 6; i < ndof; ++i) {
            v.set(i, 0.2);
        }

        // Generalized momentum p = H(q) v
        SimpleMatrix p = H.mult(v);

        System.out.println("v = ");
        v.print();
        System.out.println("p = ");
        p.print();
    }
}
      

File: Robotics_Kinematics_and_Dynamics_Course/Chapter17/Lesson1/Chapter17_Lesson1.cpp
#include <iostream>
#include <Eigen/Dense>

int main() {
  const int n_joints = 12;
  const int ndof = 6 + n_joints;

  Eigen::VectorXd q(ndof);
  Eigen::VectorXd v(ndof);
  Eigen::MatrixXd H(ndof, ndof);

  q.setZero();
  v.setZero();
  H.setIdentity(); // Example only; real H(q) from dynamics library

  // Example base twist
  v.segment<6>(0) << 0.0, 0.0, 0.5, 0.1, 0.0, 0.0;
  // Example joint velocities
  v.segment(6, n_joints).setConstant(0.2);

  // Generalized momentum
  Eigen::VectorXd p = H * v;

  std::cout << "v = " << v.transpose() << std::endl;
  std::cout << "p = " << p.transpose() << std::endl;

  return 0;
}
      
